src/lwt/win/lwt_unix_impl.ml

```diff
@@ -363,6 +367,9 @@ if !d then Format.eprintf "Start writing@.";
 if !d then Format.eprintf "Writing started@.";
   res
 
+let write ch b pos len =
+  write_substring ch (Bytes.to_string b) pos len
```

Comment:
Here, you are copying `b`, which can be costly and avoided. I would put into `write`'s body the same as `write_substring`, with the call to `blit_string_to_buffer` replaced with `blit_bytes_to_buffer`.

Reply:
I don't know how to construct `blit_bytes_to_buffer`. Can you suggest the correct code change?

Reply:
> I don't know how to construct `blit_bytes_to_buffer`.

You can construct `blit_bytes_to_buffer` by copying and adapting `blit_string_to_buffer`. For now, it is an external, and maybe just adapting the type works. But this is not future-proof. Ideally, the copying/adapting should be done with the C functions as well.

> Can you suggest the correct code change?

Are there instructions on how to reproduce the build environment locally?

Reply:
Yes, I've got a detailed recipe that I'm working into a PR for the Windows PC READMEs ...

- install Cygwin

```batch (CMD shell)
@rem # CMD shell

@rem # configuration/customization (user may alter at their discretion)
set "CYGWIN_ARCH=x86_64" &@rem # [x86,x86_64] ; x86/32-bit can be used for 32-bit build machines (if needed), but the x86_64/64-bit version can build either 32-bit or 64-bit binaries
set "CYGWIN_ROOT=c:\cygwin"
set "CYGWIN_URL=http://cygwin.mirror.constant.com"

@rem # basic set of default packages for the Cygwin/OCaml toolchain
set "PACKAGES=curl,diffutils,dos2unix,git,m4,make,mercurial,patch,perl,rsync,unzip,mingw64-i686-gcc-core,mingw64-x86_64-gcc-core"
if /i "%CYGWIN_ARCH" == "x86_64" set "PACKAGES=%PACKAGES%,mingw64-x86_64-gcc-core"

@rem # 'lablgtk' prerequisite packages
set "PACKAGES=%PACKAGES%,mingw64-i686-glib2.0,mingw64-i686-gtk2.0,mingw64-i686-pango1.0,libgtk2.0_0,libgtk2.0_devel,ocaml-lablgtk2"
if /i "%CYGWIN_ARCH" == "x86_64" set "PACKAGES=%PACKAGES%,mingw64-x86_64-glib2.0,mingw64-x86_64-gtk2.0,mingw64-x86_64-pango1.0"

@rem # install Cygwin + packages
@rem # * `setup-%CYGWIN_ARCH%.exe` can be downloaded from https://cygwin.com/install.html
setup-%CYGWIN_ARCH%.exe --no-shortcuts --quiet-mode --root "%CYGWIN_ROOT%" --site "%CYGWIN_URL%" --packages %PACKAGES%

@rem # start Cygwin bash shell
"%CYGWIN_ROOT%"\Cygwin.bat
```

- install/configure OCaml, download unison source, and build unison

```shell (Cygwin/bash)
# Cygwin/bash shell

# configuration/customization (user may alter at their discretion)
BUILDER_ARCH=64 ## [32,64] ; note: must match 'bitness' of CYGWIN_ARCH
DOWNLOADS_DIR=/tmp/ocaml.x${BUILDER_ARCH}-downloads
OCAML_VERSION=4.08.1
PROJECTS_DIR=~/Projects
TARGET_ARCH=32 ## [32,64]
TARGET_BASEDIR=${PROJECTS_DIR}
TOOLSET=mingw ## [mingw,msvc] ; note: 'MSVC' is not working because of GTK+/lablgtk installation blockers
UNISON_REPO_URL=https://github.com/bcpierce00/unison.git
UNISON_REPO_COMMIT=master
UNISON_REPO_DIR=${PROJECTS_DIR}/unison

UNISON_FETCH=https://github.com/rivy/unison fix.builds+publishing ## remove when merged into master

# * OCAML installation/setup

SWITCH=${OCAML_VERSION}+${TOOLSET}${TARGET_ARCH}

BUILDER_ARCH_ID="i686"
case "${BUILDER_ARCH}" in "64") BUILDER_ARCH_ID="x86_64" ;; esac ;
TARGET_ARCH_ID="i686"
case "${TARGET_ARCH}" in "64") TARGET_ARCH_ID="x86_64" ;; esac ;

mkdir -p "${DOWNLOADS_DIR}"
pushd "${DOWNLOADS_DIR}"
curl -#L https://github.com/fdopen/opam-repository-mingw/releases/download/0.0.0.2/opam${BUILDER_ARCH}.tar.xz -o "${DOWNLOADS_DIR}/opam${BUILDER_ARCH}.tar.xz"
tar xvf "opam${BUILDER_ARCH}.tar.xz"
bash opam${BUILDER_ARCH}/install.sh
popd

OPAM_REPOSITORY_URL=https://github.com/fdopen/opam-repository-mingw.git#opam2
opam init --disable-sandboxing --enable-completion --enable-shell-hook --auto-setup default "${OPAM_REPOSITORY_URL}"
opam update

opam switch create ${SWITCH} ${SWITCH}c ## using '...c' pre-compiled version decreases switch installation time by approximately 50%
opam switch ${SWITCH}
eval "$(opam env)"
case "${TOOLSET}" in "mingw")
  ocaml-env exec -- opam install --yes depext-cygwinports depext
  ocaml-env exec -- opam install --yes lablgtk
  ;; esac
case "${TOOLSET}" in "msvc")
  ocaml-env exec -- opam install --yes depext
  ##ocaml-env exec -- opam install --yes lablgtk ## doesn't work
  ;; esac

# * install/build `unison`

mkdir -p "${PROJECTS_DIR}"
mkdir -p "${UNISON_REPO_DIR}"
mkdir -p "${TARGET_BASEDIR}"

## * unison source
git clone "${UNISON_REPO_URL}" "${UNISON_REPO_DIR}"
pushd "${UNISON_REPO_DIR}"
git checkout ${UNISON_REPO_COMMIT} -f ## optional
if [ -n "${UNISON_FETCH}" ] ; then git fetch ${UNISON_FETCH} && git checkout FETCH_HEAD ; fi

## * build unison UISTYLE=text
VERSION_TEXT=$(git describe)
TARGET_DIR="${TARGET_BASEDIR}/unison-${VERSION_TEXT}-text-cygwin.x${BUILDER_ARCH}+ocaml-${SWITCH}"
git clean -fdx
make src OSTYPE=$OSTYPE UISTYLE=text
strip src/*.exe
mkdir -p "${TARGET_DIR}"
cp src/*.exe "${TARGET_DIR}"/

## * build unison UISTYLE=gtk2
VERSION_TEXT=$(git describe)
TARGET_DIR="${TARGET_BASEDIR}/unison-${VERSION_TEXT}-gtk2-cygwin.x${BUILDER_ARCH}+ocaml-${SWITCH}"
git clean -fdx ## needed? => test that generated executable is the same even if not cleaned after the UISTYLE=text build
make src OSTYPE=$OSTYPE UISTYLE=gtk2
strip src/*.exe
mkdir -p "${TARGET_DIR}"/bin
cp src/*.exe "${TARGET_DIR}"/bin
# * package artifacts with required GTK+ files; ref: <https://gtk-rs.org/docs-src/tutorial/cross> @@ <https://archive.is/wip/tIulF>
# [gtk-rs ~ bundling] ref: https://gtk-rs.org/docs-src/tutorial/cross
# ... + gdk-pixbuf; ref: https://github.com/gtk-rs/gtk/issues/422#issuecomment-686927113
# required DLLs
dll_refs() { eval "$(opam env)" ; eval "$(ocaml-env cygwin)" ; objdump -p "$@" | grep -Po "\S+[.]dll$" | xargs -I{} 2>/dev/null which '{}' | sort -u ; }
filtered_dll_refs() { list="$(dll_refs "$@" | grep -vF "$(cygpath ${WINDIR})" | perl -lape '$_ = qq/@{[sort @F]}/')" ; echo -n "$list" ; }
recursive_dll_refs() { list="$(dll_refs "$@")" ; n=0 ; while [ $n -lt $(echo "$list" | wc -l) ]; do n=$(echo "$list" | wc -l) ; list="$(dll_refs $list)" ; done ; echo -n "$list" ; }
recursive_filtered_dll_refs() { list="$(filtered_dll_refs "$@")" ; n=0 ; while [ $n -lt $(echo -n "$list" | wc -l) ]; do n=$(echo -n "$list" | wc -l) ; list="$(filtered_dll_refs $list)" ; done ; echo -n "$list" ; }
IFS=$'\n' DLL_list=( "$(recursive_filtered_dll_refs "${TARGET_DIR}"/bin/*.exe)" )
for dll in ${DLL_list[@]} ; do cp "${dll}" "${TARGET_DIR}"/bin ; done
mkdir -p "${TARGET_DIR}"/lib
# * copy gdk-pixbuf lib/dlls
cp -r /usr/${TARGET_ARCH_ID}-w64-mingw32/sys-root/mingw/lib/gdk-pixbuf-2.0 "${TARGET_DIR}"/lib/
# * update loader.cache to point to local relative installation
mv "${TARGET_DIR}"/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache "${TARGET_DIR}"/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache.original
cat "${TARGET_DIR}"/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache.original | sed -E 's#([^"]*)(lib/gdk-pixbuf-2.0/2.10.0/loaders/[^"]*[.]dll)#../\2#' > "${TARGET_DIR}"/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache

popd
```

Reply:
The scripts don't work totally out of the box, but I managed to do something. I've pushed to your branch what I had in mind.

Reply:
Can you give me feedback on what problems you encountered?